resilience4j-spring/src/main/java/io/github/resilience4j/circuitbreaker/configure/CircuitBreakerAnnotationConfigScanner.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2019 Olov Andersson
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.github.resilience4j.circuitbreaker.configure;
+
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.config.BeanDefinition;
+import org.springframework.beans.factory.config.BeanFactoryPostProcessor;
+import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
+
+import io.github.resilience4j.circuitbreaker.annotation.CircuitBreaker;
+import io.github.resilience4j.common.circuitbreaker.configuration.CircuitBreakerConfigurationProperties.InstanceProperties;
+
+/**
+ * Scans all bean definitions for CircuitBreaker annotations and collects
+ * annotated configuration so it can be merged with the properties
+ * configuration. This is done as a BeanFactoryPostProcessor so we can ensure
+ * that we have found these before actual bean instantiation begins.
+ * <p>
+ * Unless resilience4j-spring-boot/resilience4j-spring-boot2/
+ * resilience4j-spring-cloud/resilience4j-spring-cloud2 is used this class
+ * must be explicitly added to the spring context as a bean for CircuitBreaker
+ * annotation exception config to be considered.
+ */
+public class CircuitBreakerAnnotationConfigScanner implements BeanFactoryPostProcessor {
+	private static final Logger LOG = LoggerFactory.getLogger(CircuitBreakerAnnotationConfigScanner.class);
+
+	private Map<String, InstanceProperties> instanceProperties = new HashMap<>();
+
+	@Override
+	public void postProcessBeanFactory(ConfigurableListableBeanFactory beanFactory) throws BeansException {
+		for (String beanName : beanFactory.getBeanDefinitionNames()) {
+			BeanDefinition beanDefinition = beanFactory.getBeanDefinition(beanName);
+			if (beanDefinition != null) {
+				String beanClassName = beanDefinition.getBeanClassName();
+				if (beanClassName != null) {
+					try {
+						Class<?> beanClass = Class.forName(beanClassName);
+						findAnnotationConfig(beanClass);
+					} catch (ClassNotFoundException e) {
+						LOG.error("Failed to load beanClass {} while searching for CircuitBreaker annotations", e);
+					}
+				}
+			}
+
+		}
+	}
+
+	private void findAnnotationConfig(Class<?> beanClass) {
+		for (Method m : beanClass.getMethods()) {
```

Comment:
I hadn't time to look at the PR yet. I just wonder how do we solve the case when there are multiple annotations for the same CircuitBreaker instance, but with different exceptions configured?

Reply:
@Romeh Modified it to consider class-level annotations as well.

@RobWin The only reasonable behavior for conflicting annotation configurations would be to regard them as a programming error and throw an exception, so I implemented that. When it comes to external properties it is however consistent with spring in general to consider them as overrides.

Reply:
For the multiple annotations case in the same class , i guess we need to merge the configured exceptions from multiple annotations for the same instance , as the user maybe would like to define common ignored exception over the class itself the he can define more specific exceptions to be ignored over the methods itself , example of what i mean : 


```
@CircuitBreaker(name = DummyService.BACKEND,recordExceptions = {IOException.class})
@Component
public class DummyServiceImpl {

	@CircuitBreaker(name = DummyService.BACKEND,recordExceptions = {BussinessExcp.class})
	public void doSomething(boolean throwBackendTrouble){
	}
}
```

@RobWin  what do u think ?

Reply:
I think that is reasonable within a class, but I don't think it should be allowed between different classes, as that probably is more likely to be a programming error. Should we search the class hierarchy as well?

Reply:
I more and more dislike the option to configure anything via annotations. I becomes too complicated to understand.  But I do understand that compile-safety is worth the effort. 😢

I don't think that we should allow the inheritance of annotations. Every annotation should stay for it's own.

Reply:
what about interfaces ? like Feign client usage ? would be covered as well in annotation config support  ?

Reply:
True dat. I guess Spring creates a proxy bean for the interface and the annotation is not at class, but at interface-level.

Reply:
As you say, we really need to take the interfaces into consideration and we should probably include super-classes as well to be consistent. But although, this might appear a bit complicated, it's no different than what Spring does on a regular basis for lots of things (Spring must do it to build the actual circuit breaker aspects for instance). Spring has a utility class to help finding the annotations in these cases: AnnotationUtils.

And on the upside, the logic is concentrated to the single CircuitBreakerAnnotationConfigScanner class and these kind of things are very easy to cover and verify with unit tests.

Reply:
@RobWin @Romeh I pushed a fix to search the class hierarchy for annotations, but the build failed a unit test in an entirely unrelated module. I discovered that it builds correctly locally most of the times, bit failed at one point. Seem to me like there are some unit tests that might fail because of timing issues? Is there a way to trigger the build again?

Also the thing we discussed above that complicates things would be if we would try to merge annotation configuration for annotations. For instance should the exception list be merged or replaced. I think it might be easier to just not allow it as it is now. That would add a really useful feature, without making it too complex to understand. Property file vs annotation configuration is easier to reason about, the code is the default and the property file is the override, which replaces the code (since it could be re-configured without re-compilation and for spring-cloud, it could even be re-configured in runtime).

Reply:
Yes I agree. We should not merge annotations.

Reply:
But what does that mean? First discovered annotation wins? Or last discovered annotation wins?

Reply:
Yes, currently the code throws an IllegalArgumentException in this case (there's a junit test case for this).

Reply:
Note, that it will only throw an exception if there's an actual conflict. So you're allowed to re-use the annotation, as long as you don't define any conflicting exceptions.

Reply:
It should be possible to use exactly the same Annotation in different classes. Because it's possible at the moment.
If you use `@CircuitBreaker(name = DummyService.BACKEND,recordExceptions = {BussinessExcp.class})` in two classes or two methods, it should be possible.

Reply:
Would it be possible to add the discovered properties from annotations to Spring's environment? Like any other `PropertySource`? Even before the CircuitBreakerConfigurationProperties are loaded?

Like: `environment.getPropertySources().addLast(new MapPropertySource("annotationBla", new HashMap<>()));`

Reply:
That way we would be able to add real properties via annotations to the environment?

Reply:
I think you can inject an `ConfigurableEnvironment environment` into the `BeanFactoryPostProcessor `

Reply:
Interesting idea, I'll look into it.

Reply:
@RobWin The PropertySource/ConfigurableEnvironment worked as expected, so I've pushed that change now. Also fixed so non-conflicting annotations (that declare the same exceptions for the same name) are allowed. Tested it in the resilience4j-spring-cloud2-demo project as well.

Question: It's a bit unclear to me why there are both CIrcuitBreakerConfigurationProperties.getConfigs() and CircuitBreakerConfigurationProperties.getInstances()?

The annotations only set the instances properties at this point.

Reply:
I guess it's a hack and the Spring Boot guys will hate us for it ;)

Reply:
`getConfigs` returns a list of Configs was can be shared between instances.

For example:

```
resilience4j.circuitbreaker:
    configs:
        default:
            slidingWindowSize: 100
            permittedNumberOfCallsInHalfOpenState: 10
            waitDurationInOpenState: 10000
            failureRateThreshold: 60
            eventConsumerBufferSize: 10
            registerHealthIndicator: true
        someShared:
            slidingWindowSize: 50
            permittedNumberOfCallsInHalfOpenState: 10
    instances:
        backendA:
            baseConfig: default
            waitDurationInOpenState: 5000
        backendB:
            baseConfig: someShared
```